### PR TITLE
Range can be accepted in Sum

### LINKED ISSUE #16806: summation doesn't accepts Range as limits
#### Problem
As said in the title I cannot pass `Range` as limits in the `summation`. However, `integrate` accepts the analogous `Interval` as limits. 

#### Example of problem
```
Python 3.6.7 (default, Oct 22 2018, 11:32:17) 
[GCC 8.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> limits = Range(1, 50)
>>> i = symbols('i', positive=True, integer=True)
>>> summation(i, (i, limits))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/gagandeep/sympy/sympy/concrete/summations.py", line 847, in summation
    return Sum(f, *symbols, **kwargs).doit(deep=False)
  File "/home/gagandeep/sympy/sympy/concrete/summations.py", line 161, in __new__
    obj = AddWithLimits.__new__(cls, function, *symbols, **assumptions)
  File "/home/gagandeep/sympy/sympy/concrete/expr_with_limits.py", line 356, in __new__
    pre = _common_new(cls, function, *symbols, **assumptions)
  File "/home/gagandeep/sympy/sympy/concrete/expr_with_limits.py", line 38, in _common_new
    limits, orientation = _process_limits(*symbols)
  File "/home/gagandeep/sympy/sympy/concrete/expr_with_limits.py", line 123, in _process_limits
    raise ValueError('Invalid limits given: %s' % str(symbols))
ValueError: Invalid limits given: ((i, Range(1, 50, 1)),)
>>> x = symbols('x', real=True, positive=True)
>>> limits0 = Interval(3, 8)
>>> integrate(x, (x, limits0))
55/2
```

#### Other comments/references
I am working on https://github.com/sympy/sympy/pull/16576 to add certain multivariate distributions. I observed that `Range` wasn't being accepted by the summation  but the sets of discrete distributions are usually described as either `Range` or pre-defined sets, like, `S.Naturals`. I have to make changes as in [this line of code](https://github.com/sympy/sympy/pull/16576/files#diff-f5bd154f958178a67c12702770445325R104). However, I think, this should have been handled by summation itself.

### PR DESCRIPTION
**Pull request by czgdp1807**
#### References to other Issues or PRs

Fixes https://github.com/sympy/sympy/issues/16806

#### Brief description of what is fixed or changed
Read the discussion at the above Issue and see the diff.

#### Other comments
N/A

#### Release Notes

* concrete
  * `Range` can be passed as limit to `Sum`

### CODE REVIEW

**Comment by czgdp1807** [comment]
@smichr I have tried to fix it, can you please take a look and tell me whether it is merge-able.

File: sympy/concrete/expr_with_limits.py [hunk 1]
```diff
@@ -90,7 +91,13 @@ def _process_limits(*symbols):
                 limits.append(Tuple(V))
             continue
         elif is_sequence(V, Tuple):
-            V = sympify(flatten(V))
```
Comment (smichr) [inline sympy/concrete/expr_with_limits.py]: You can leave this here: it will flatten Range into 3 elements (start, stop, step).

Comment (czgdp1807) [reply sympy/concrete/expr_with_limits.py]: Ah! that modification wasn't needed. I could have used the `(i, start, stop, step)` and converted it to `(i, start, stop)` if step is 1. I will make the change. I am surprised how can I miss that. 😅

File: sympy/concrete/expr_with_limits.py [hunk 2]
```diff
@@ -90,7 +91,13 @@ def _process_limits(*symbols):
...
+            new_V = V
+            if len(V) > 1:
+                if isinstance(V[1], Range):
+                    if V[1].step != 1:
+                        raise ValueError("The step of Range must be 1.")
+                    new_V = (V[0], V[1].inf, V[1].sup)
+            V = sympify(flatten(new_V))
             if isinstance(V[0], (Symbol, Idx)) or getattr(V[0], '_diff_wrt', False):
                 newsymbol = V[0]
                 if len(V) == 2 and isinstance(V[1], Interval):
```
Comment (smichr) [inline sympy/concrete/expr_with_limits.py]: add here,
```python
if len(V) == 4:
    V[2:] = 0, (V[2] - V[1])//V[3] - 1, V[0]*V[3] + v[1]  # x, Range(1, 11, 2) -flat-> x, 1, 11, 2 -> x, 0, 4, 2*x + 1
```
Then back in the line after the call to `_process_limits` could appear:
```python
for i, lim in enumerate(limits):
    if len(lim) == 4:
        x, _, _, xnew = lim
        function = function.subs(x, xnew)
        limits[i].pop()  # no need xnew
```

Comment (czgdp1807) [reply sympy/concrete/expr_with_limits.py]: @smichr I have made the modifications according to your reviews. Thanks for the suggestions.

**Comment by smichr** [comment]
Looks good. One mod that can be made in future prs is to remove the sympify of the Tuple V flatten result since all Tuple args are Basic.

**Comment by smichr** [comment]
Let's make this canonical, however, by using lo=inf hi=sup and d=abs of step. Range(9,0,-2) should give same symbolic sum of i before doit as that of Range(1,10,2).

**Comment by czgdp1807** [comment]
Thank you @smichr for making the commits. :-)

**Comment by smichr** [comment]
Hmm. That code is a little contorted. It should now work if you pass an Idx with Range. I'm not sure what should happen if lo and hi are not given with Idx, however, so am raising an error.

**Comment by czgdp1807** [comment]
Sure, we can add a TODO task for that case in the code.

**Comment by smichr** [comment]
One more time...

**Comment by smichr** [comment]
Do you have any popcorn? ;-)

**Comment by czgdp1807** [comment]
Up for a movie then? 😉

**Comment by czgdp1807** [comment]
@smichr Tests passed. Is there anything else to change?

**Comment by smichr** [comment]
No, LOL!

**Pull request merged**